### ISSUE TITLE
what is the Optimizer

### ISSUE THREAD

**Issue by cissoidx** [google-research/vision_transformer#94]
Hi, in `momentum_clip`, there defined an optimizer that stores state using half-precision. I guess it is for quantification reasons. But this optimizer is nothing like what i am familiar with. and in `flax.optim`, there are actually a lot of 'standard' optimizers. 

My question is: why use this specific optimizer and not the standard ones, any advantages? What is this optimizer actually doing, how does it update weights,any equations? any papers?

looking forward to your reply

**Comment by andsteing** [comment]
Hi,

`momentum_clip` update rule is simply momentum:

https://github.com/google-research/vision_transformer/blob/a51dbfa231262ffd1e8e52c041324e6de644d99b/vit_jax/momentum_clip.py#L69-L76

This is identical to the Flax implementation (with `weight_decay=0` and `nesterov=False`).

Differences are:

1. You can specify dtype of momentum with `--optim_dtype` flag. Using `bfloat16` has better range than `float16` but uses only 50% of memory of `float32`.

https://github.com/google-research/vision_transformer/blob/a51dbfa231262ffd1e8e52c041324e6de644d99b/vit_jax/flags.py#L68-L72

2. We also implemented `grad_norm_clip` in the optimizer - but that could also be implemented in main training loop, it's simply a matter of style.

Best, Andreas

**Comment by cissoidx** [comment]
Is there any specific reason that you prefer not to use the standard optimizers? Like SGD, ADAM?

**Comment by andsteing** [comment]
Momentum is a fairly standard optimizer as well (e.g. [flax.optim.Momentum](https://github.com/google/flax/blob/master/flax/optim/momentum.py)). As for motivation of the two differences (dtype, grad_norm_clip) see my last post.

**Comment by cissoidx** [comment]
Thanks, clearly explained.
cheers.

**Comment by cissoidx** [comment]
> Momentum is a fairly standard optimizer as well (e.g. [flax.optim.Momentum](https://github.com/google/flax/blob/master/flax/optim/momentum.py)). As for motivation of the two differences (dtype, grad_norm_clip) see my last post.

As I can see that momentum datatype can be specified in your code, then what is the datatype for weights and gradients?

**Comment by andsteing** [comment]
Weights and gradients are the default `float32`. You can inspect the `dtype` of every state e.g. by

```
jax.tree_map(lambda x: x.dtype, flax.serialization.to_state_dict(opt_repl))
```

(add this in a new cell after https://colab.sandbox.google.com/github/google-research/vision_transformer/blob/master/vit_jax.ipynb#scrollTo=zKn4IfUWHWPk )

So yes, this is a sort of mixed precision training, that allowed us to train larger models on TPU. While it reproduces the same numbers on GPU, the benefits are probably somewhat less (since `bfloat16` is not a native datatype).

If you're interested in mixed precision training using JAX on GPU, have a look at the `imagenet` example in the main Flax repo:

https://github.com/google/flax/blob/master/examples/imagenet/train_test.py

Note that in general mixed precision training is very sensitive to the exact application (optimizer, domain, model, ...) and there is no one-size-fits-all solution. I would recommend in the general case to start with full precision and only experimenting with mixed precision once you get good results with the baseline model.
